presto-tests/src/main/java/com/facebook/presto/tests/QueryAssertions.java

```diff
@@ -351,6 +351,82 @@ public static void copyTpchTables(
         log.info("Loading from %s.%s complete in %s", sourceCatalog, sourceSchema, nanosSince(startTime).toString(SECONDS));
     }
 
+    public static void copyTpcdsTables(
```

Comment:
I see there are many lines of code added but a lot of them are doing the same thing. I'm considering a refactoring before introducing the TPCDS logic.

The code for copying non-bucketed tables in `copyTpchTables` and  `copyTpcdsTables` are actually not relevant to which catalog it is. They're both just constructing a copy-table sql and execute it for every table. Also the sourceCatalog is already in the method parameter list so it can be easily read and understood. There is no need to have two separate methods or even more in the future.

Further more, the main loop for copying of bucketed tables are the same as copying non-bucketed tables, the only difference is the SQL for different tables are different. I also noticed that `copyTableBucketed` for TPCDS is in QueryAssertions.java, while the one for TPCH is in HiveQueryRunner.java. It'll be better to put them together.

How about we rewrite them like this:

In QueryAssertions.java

```
public static void copyTables(
            QueryRunner queryRunner,
            String sourceCatalog,
            String sourceSchema,
            Session session,
            Iterable<String> tables,
            boolean ifNotExists,
            boolean bucketed)
    {
        log.info("Loading data from %s.%s...", sourceCatalog, sourceSchema);
        long startTime = System.nanoTime();
        for (String table : tables) {
            copyTable(queryRunner, sourceCatalog, sourceSchema, table, session, ifNotExists, bucketed);
        }
        log.info("Loading from %s.%s complete in %s", sourceCatalog, sourceSchema, nanosSince(startTime).toString(SECONDS));
    }

private static void copyTable(QueryRunner queryRunner, String sourceCatalog, String sourceSchema, String sourceTable, Session session, boolean ifNotExists, boolean bucketed)
    {
        QualifiedObjectName table = new QualifiedObjectName(sourceCatalog, sourceSchema, sourceTable);
        long start = System.nanoTime();
        log.info("Running import for %s", table.getObjectName());

        @Language("SQL") String sql = getCopyTableSql(sourceCatalog, table, ifNotExists, bucketed);
        long rows = (Long) queryRunner.execute(session, sql).getMaterializedRows().get(0).getField(0);

        log.info("Imported %s rows for %s in %s", rows, table.getObjectName(), nanosSince(start).convertToMostSuccinctTimeUnit());

 private   static String getCopyTableSql(String sourceCatalog, QualifiedObjectName table, boolean ifNotExists, boolean bucketed) {
        if (!bucketed) {
            return format("CREATE TABLE %s %s AS SELECT * FROM %s", ifNotExists ? "IF NOT EXISTS" : "", table.getObjectName(), table);
        }
        else {
            switch (sourceCatalog) {
                case "tpch" : return getTpchCopyTableSqlBucketed(table);
                case "tpcds" : return getTpcdsCopyTableSqlBucketed(table);
                default: throw new UnsupportedOperationException();
            }
        }
    }

    private static String getTpchCopyTableSqlBucketed(QualifiedObjectName table)
    {
        @Language("SQL") String sql;
        switch (table.getObjectName()) {
            case "part":
            case "partsupp":
            case "supplier":
            case "nation":
            case "region":
                sql = format("CREATE TABLE %s AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "lineitem":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['orderkey'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "customer":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['custkey'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "orders":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['custkey'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            default:
                throw new UnsupportedOperationException();
        }
        return sql;
    }

    private static String getTpcdsCopyTableSqlBucketed(QualifiedObjectName table)
    {
        @Language("SQL") String sql;
        switch (table.getObjectName()) {
            case "call_center":
            case "catalog_page":
            case "customer":
            case "customer_address":
            case "customer_demographics":
            case "date_dim":
            case "household_demographics":
            case "income_band":
            case "item":
            case "promotion":
            case "reason":
            case "ship_mode":
            case "store":
            case "time_dim":
            case "warehouse":
            case "web_page":
            case "web_site":
                sql = format("CREATE TABLE %s AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "inventory":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['inv_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "store_returns":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['sr_returned_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "store_sales":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['ss_sold_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "web_returns":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['wr_returned_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "web_sales":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['ws_sold_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "catalog_returns":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['cr_returned_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            case "catalog_sales":
                sql = format("CREATE TABLE %s WITH (bucketed_by=array['cs_sold_date_sk'], bucket_count=11) AS SELECT * FROM %s", table.getObjectName(), table);
                break;
            default:
                throw new UnsupportedOperationException();
        }

        return sql;
    }
```
And remove the `copyTpchTables` and `copyTpcdsTables` methods, and `private static void copyTable(QueryRunner queryRunner, QualifiedObjectName table, Session session, boolean ifNotExists)` method.

In HiveQueryRunner.java, remove `copyTpchTablesBucketed()` and `copyTableBucketed()`., and call the new methods like this:
```
copyTables(queryRunner, "tpch", TINY_SCHEMA_NAME, createSession(Optional.empty()), transform(tables, table -> table.getTableName()), true, false);

copyTables(queryRunner, "tpcds", TINY_SCHEMA_NAME, createSession(Optional.empty()), tpcdsTables, true, false);
```

The code will look shorter in this refactoring.

Reply:
agree. I'm working on the refactor to make it more readable for both tpch and tpcds as you pointed out. many thanks for sample code as reference! Will push out a new commit to review.